### DISCUSSION TITLE
"after_flush" listener added in "after_insert" listener

### DISCUSSION THREAD

**Discussion by thibault-yellowstones** [sqlalchemy/sqlalchemy#10449] [Usage Questions] [answered]
Hi community!
I am trying to fix an odd bug and would love to understand with certainty what is happening in our SQLAlchemy code. here it is:
```py
@event.listens_for(TargetModel, "after_insert")
def receive_after_insert(mapper, connection, target):
    @event.listens_for(DBSession, "after_flush", once=True)
    def receive_after_flush(session, flush_context):
         target.add_log_after_insert()
```

Sometimes, inserting the logs raises an `IntegrityError`, because the target does not exist. My understanding is that the target must have been inserted then removed from the session at the time the log is inserted.

First, any idea why there is a double listener here? Seems like just the first layer would be fine? Someone told me it is to avoid this warning:
> [SAWarning: Usage of the 'Session.add()' operation is not currently supported within the execution stage](https://stackoverflow.com/questions/43164546/sawarning-usage-of-the-session-add-operation-is-not-currently-supported-wit)

Second, what is the best way to fix the bug? I am tempted to add a `if target in session.new` condition before the log insert but that is purely because I don't want to change a code that isn't clear enough...

Thanks!

**Comment by zzzeek** [comment, marked as answer]
the above code will add many new event listeners to the DBSession class over and over again, every time a TargetModel is inserted, which will cause extremely poor results and also leak memory since there can be infinite TargetModel inserts.   you should not be adding an event listener in this way.

if you want to see what objects were inserted in the after_flush event, you can iterate through the `Session.new` collection, or if you want to coordinate between two events like this, use a data structure on the Session:


```py

@event.listens_for(TargetModel, "after_insert")
def receive_after_insert(mapper, connection, target):
    session = inspect(target).session
    if "inserted_target_models" not in session.info:
       session.info["inserted_target_models"] = set()
    session.info["inserted_target_models"].add(target)

@event.listens_for(DBSession, "after_flush", once=True)
def receive_after_flush(session, flush_context):
      for target in session.info.get("inserted_target_models", ()):
          target.add_log_after_insert()
      session.info.pop("inserted_target_models")
```

**Reply by thibault-yellowstones** [reply]
Thanks @zzzeek! 
Isn't the point of `once` to register a listener to fire only once though? I was imagining that the listener was discarded after the first call and the performances were not that bad.
Regardless, I agree your option looks much more effective 👍

**Reply by thibault-yellowstones** [reply]
Oh I just saw the warning in the doc 🙈about de-registration.

**Reply by CaselIT** [reply]
@zzzeek maybe regarding de-registration of once events, instead of complicating things while the event is run, we could remove the ones that are done while registering a new event?
Ideally event registration should be less time critical

**Reply by zzzeek** [reply]
talking about once=True?    I recall there being big problems getting once=True to do things more cleanly, feel free to look into it though

**Reply by CaselIT** [reply]
IIRC the main issue is that it can't be removed after calling it since it's in a loop then and items can't be remove from a collection while looping it, and working around that issue would cause quite a few complication on a performance critical section.

maybe doing maintenance on add is easier?

**Reply by tcowpop** [reply]
@zzzeek regarding having a `@event.listens_for(DBSession, "after_flush", once=True)`. Would you only have to have a single one for all models?

And then one of these per Model?:

```
@event.listens_for(TargetModel, "after_insert")
```

Thanks!

**Reply by zzzeek** [reply]
after_flush applies to the Session target, and after_insert applies to individual model targets, so that would be correct.  you can apply after_insert to your root `Base` class or any superclass and add the parameter `propagate=True` where it will take effect for all subclasses.
